edb/edgeql/compiler/inference/multiplicity.py

```diff
@@ -0,0 +1,701 @@
+            if (expr_mult is MANY and
+                    not ptrref.out_target.is_scalar and
+                    not ptrref.out_target.collection):
+                raise errors.QueryError(
+                    f'possibly not a strict set returned by an '
+                    f'expression for a computable '
+                    f'{ptrref.shortname.name}.',
+                    hint=(
+                        f'Use DISTINCT for the entire computable expression '
+                        f'to resolve this.'
+                    ),
+                    context=shape_set.context
+                )
+
+        _infer_shape(
+            shape_set, is_mutation=is_mutation, scope_tree=scope_tree, ctx=ctx)
+
+
+def _infer_set(
+    ir: irast.Set,
+    *,
+    is_mutation: bool=False,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    result = _infer_set_inner(
+        ir, is_mutation=is_mutation, scope_tree=scope_tree, ctx=ctx)
+    ctx.inferred_multiplicity[ir, scope_tree] = result
+    # The shape doesn't affect multiplicity, but requires validation.
+    _infer_shape(ir, is_mutation=is_mutation, scope_tree=scope_tree, ctx=ctx)
+
+    return result
+
+
+def _infer_set_inner(
+    ir: irast.Set,
+    *,
+    is_mutation: bool=False,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    rptr = ir.rptr
+    new_scope = cardinality._get_set_scope(ir, scope_tree)
+
+    if rptr is not None:
+        # Validate the source regardless
+        infer_multiplicity(rptr.source, scope_tree=new_scope, ctx=ctx)
+
+    if ir.expr:
+        expr_mult = infer_multiplicity(ir.expr, scope_tree=new_scope, ctx=ctx)
+
+    if rptr is not None:
+        rptrref = rptr.ptrref
+
+        # Validate the source
+        infer_multiplicity(rptr.source, scope_tree=new_scope, ctx=ctx)
+
+        if isinstance(rptr.ptrref, irast.TupleIndirectionPointerRef):
+            # All bets are off for tuple elements.
+            return MANY
+        elif ir.typeref.is_scalar or ir.typeref.collection:
+            # This is not an expression and is some kind of scalar, so
+            # multiplicity cannot be guaranteed to be ONE (most scalar
+            # expressions don't have an implicit requirement to be sets)
+            # unless we also have an exclusive constraint.
+
+            if rptr is not None:
+                schema = ctx.env.schema
+                exclusive_constr = schema.get(
+                    'std::exclusive', type=s_constr.Constraint)
+                # We should only have some kind of path terminating in a
+                # property here.
+                assert isinstance(rptrref, irast.PointerRef)
+                _ptr = schema.get_by_id(rptrref.id)
+                assert isinstance(_ptr, s_pointers.Pointer)
+                ptr = _ptr.get_nearest_non_derived_parent(ctx.env.schema)
+                is_unique = (
```

Comment:
There is actually a `Pointer.is_exclusive()` helper for this.

Reply:
Updated.